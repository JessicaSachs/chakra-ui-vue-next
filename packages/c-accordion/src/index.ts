import { h, defineComponent, PropType } from 'vue'
import { chakra, DOMElements } from '@chakra-ui/vue-system'

const CAccordion = defineComponent({
  props: {
    as: {
      type: [String] as PropType<DOMElements>,
      default: 'div',
    },
  },
  setup(props, { slots, attrs }) {
    return () => h(chakra(props.as), { ...attrs }, slots)
  },
})

export default CAccordion
