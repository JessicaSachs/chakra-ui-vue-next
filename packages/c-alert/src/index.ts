import { h, defineComponent, Component } from 'vue'

const CAlert = defineComponent({
  props: {
    as: {
      type: [String, Object],
      default: 'div',
    },
  },
  render() {
    return h(
      this.as,
      {
        ...this.$attrs,
        role: 'alert',
      },
      this.$slots.default && this.$slots.default()
    )
  },
})

export default CAlert
