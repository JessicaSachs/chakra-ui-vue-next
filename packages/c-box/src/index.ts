import { h, defineComponent, PropType } from 'vue'

const CBox = defineComponent({
  props: {
    as: {
      type: String as PropType<string>,
      default: 'div',
    },
  },
  render() {
    return h(this?.as, { ...this.$props, ...this.$attrs }, this.$slots)
  },
})

export default CBox
