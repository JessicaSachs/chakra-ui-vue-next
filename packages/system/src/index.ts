export * from './chakra'
export * from './system.attrs'
export * from './system.types'
export * from './system.utils'
