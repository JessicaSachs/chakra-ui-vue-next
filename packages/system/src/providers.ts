import { createContext, Dict } from '@chakra-ui/vue-utils'
import { SystemStyleObject } from '@chakra-ui/styled-system'

const [StylesProvider, useStyles] = createContext<Dict<SystemStyleObject>>({
  name: 'StylesContext',
  errorMessage:
    'useStyles: `styles` is undefined. Seems you forgot to provide `StylesProvider(...)` ',
})

export { StylesProvider, useStyles }
