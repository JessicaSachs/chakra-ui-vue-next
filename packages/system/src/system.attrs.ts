import kebabCase from 'lodash.kebabcase'
import camelCase from 'lodash.camelcase'
import {
  propNames,
  pseudoPropNames,
  StyleObjectOrFn,
} from '@chakra-ui/styled-system'
import { HTMLAttributes } from 'vue'

type StyleAndHTMLAttibutes = StyleObjectOrFn &
  Record<string, string | number | boolean | unknown> &
  HTMLAttributes

export const baseStylePropNames = propNames.reduce(
  (acc: StyleAndHTMLAttibutes, curr: string) => {
    acc[curr] = true
    acc[kebabCase(curr)] = true
    return acc
  },
  {}
)

export const pseudoStylePropNames = pseudoPropNames.reduce(
  (acc: StyleAndHTMLAttibutes, curr: string) => {
    acc[curr] = curr
    return acc
  },
  {}
)

export const allStylePropNames: StyleAndHTMLAttibutes = {
  ...baseStylePropNames,
  ...pseudoStylePropNames,
}

interface ExtractedStyleAttrs {
  styles: Partial<StyleAndHTMLAttibutes>
  attrs: Partial<HTMLAttributes>
}

/** Extracts CSS style properties and HTML attributes from merged component attributs */
export const extractStyleAttrs = <
  T extends StyleAndHTMLAttibutes,
  U extends Partial<HTMLAttributes> | Record<string, string | number | boolean>
>(
  styleProps: T
): ExtractedStyleAttrs => {
  const styles = {} as T
  const attrs = {} as U

  for (const prop in styleProps) {
    // TODO: Cache style prop names so that that camelCases is only computed once for each property
    // TODO: Is there another way to detect
    const _attr = (pseudoStylePropNames[prop] as string) || camelCase(prop)
    if (_attr in allStylePropNames) {
      // @ts-expect-error Not sure how to cast returned string into typeof key of U
      styles[_attr] = styleProps[prop]
    } else {
      // @ts-expect-error Not sure how to cast returned string into typeof key of U
      attrs[prop] = styleProps[prop]
    }
  }

  return {
    styles,
    attrs,
  }
}
