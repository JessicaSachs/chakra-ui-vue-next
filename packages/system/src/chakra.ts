import { DefineComponent, defineComponent, h } from 'vue'
import { css } from '@chakra-ui/styled-system'
import theme from '@chakra-ui/vue-theme'
import { cx } from '@chakra-ui/vue-utils'
import { css as _css } from '@emotion/css'
import { extractStyleAttrs } from './system.attrs'
import { domElements, DOMElements } from './system.utils'

/**
 * Creates a Chakra UI Vue component
 * @param tag Tag
 * @param componentName Component name
 */
// @ts-expect-error
export const chakra: IChakraFactory = (
  tag: DOMElements,
  componentName?: string
): DefineComponent => {
  return defineComponent({
    inheritAttrs: false,
    setup(props, { slots, attrs }) {
      // Separate component style attributes from raw HTML attributes
      const { class: inheritedClass, ...rest } = attrs
      const { styles, attrs: _attrs } = extractStyleAttrs(rest)
      const className = _css(css(styles)({ theme }))

      const _componentName = componentName ? `chakra-${componentName}` : ''

      return () =>
        h(
          tag,
          {
            class: cx(inheritedClass, _componentName, className),
            ...props,
            ..._attrs,
          },
          slots
        )
    },
  })
}

type IChakraFactory = {
  [key in DOMElements]: DefineComponent
} & {
  (tag: DOMElements, componentName?: string): DefineComponent
}

domElements.forEach((tag) => {
  chakra[tag] = chakra(tag)
})

export { domElements }
