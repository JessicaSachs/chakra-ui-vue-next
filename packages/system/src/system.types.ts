import {
  Component,
  Fragment,
  Suspense,
  Teleport,
  ComponentObjectPropsOptions,
} from 'vue'
import {
  SystemProps,
  ResponsiveValue,
  StyleProps,
} from '@chakra-ui/styled-system'
import { DOMElements } from './system.utils'
import { IntrinsicElementAttributes } from './dom.types'

export type Tag =
  | string
  | typeof Fragment
  | typeof Teleport
  | typeof Suspense
  | Component

export interface ThemingProps {
  variant?: string
  size?: string
  colorScheme?: string
  orientation?: 'vertical' | 'horizontal'
  styleConfig?: any
}

export interface ChakraProps extends SystemProps {
  /**
   * apply layer styles defined in `theme.layerStyles`
   */
  layerStyle?: string
  /**
   * apply typography styles defined in `theme.textStyles`
   */
  textStyle?: string
  /**
   * Reference styles from any component or key in the theme.
   *
   * @example
   * ```html
   * <c-box apply="styles.h3">This is a div</c-box>
   * ```
   *
   * This will apply styles defined in `theme.styles.h3`
   */
  apply?: string
  /**
   * if `true`, it'll render an ellipsis when the text exceeds the width of the viewport or maxWidth set.
   */
  isTruncated?: boolean
  /**
   * Used to truncate text at a specific number of lines
   */
  noOfLines?: ResponsiveValue<number>
}

type ElementType<P = any> =
  | {
      [K in keyof IntrinsicElementAttributes]: P extends IntrinsicElementAttributes[K]
        ? K
        : never
    }[keyof IntrinsicElementAttributes]
  | Component<P>

export type As<Props = any> = ElementType<Props>

/**
 * Extract the props of a Vue element or component
 */
export type PropsOf<T extends As> = ComponentObjectPropsOptions<T> & {
  as?: As
}

export type HTMLChakraProps<T extends As> = Omit<
  PropsOf<T>,
  T extends 'svg'
    ? 'ref' | 'children' | keyof StyleProps
    : 'ref' | keyof StyleProps
> &
  ChakraProps & { as?: As }
