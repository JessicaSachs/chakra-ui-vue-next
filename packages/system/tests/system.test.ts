import { render, screen } from '@chakra-ui/vue-test-utils'
import { defineComponent, h } from 'vue'
import { chakra } from '../src'

const renderComponent = (options?: Record<string, any>) =>
  render({
    components: {
      chakra: defineComponent({
        setup(_, { slots }) {
          return () => h(chakra('span'), {}, slots)
        },
      }),
    },
    template: `<chakra>child-element</chakra>`,
    ...options,
  })

it('should be render properly', () => {
  const { asFragment } = renderComponent()
  expect(asFragment()).toMatchSnapshot()
})

it('should be render default slot', () => {
  renderComponent()
  expect(screen.getByText('child-element')).toBeInTheDocument()
})
