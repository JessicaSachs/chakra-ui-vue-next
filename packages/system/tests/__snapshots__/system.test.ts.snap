// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be render properly 1`] = `
<DocumentFragment>
  <span
    class="css-0"
  >
    child-element
  </span>
</DocumentFragment>
`;
